deref patterns: implement implicit deref patterns

dianne opened the pull request:
This implements implicit deref patterns (per https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg#Implicit-deref-patterns) and adds tests and an unstable book chapter.

Best reviewed commit-by-commit. Overall there's a lot of additions, but a lot of that is tests, documentation, and simple(?) refactoring.

Tracking issue: #87121

r? @Nadrieril

dianne committed:
lower implicit deref patterns to THIR

Since this uses `pat_adjustments`, I've also tweaked the documentation
to mention implicit deref patterns and made sure the pattern migration
diagnostic logic accounts for it. I'll adjust `ExprUseVisitor` in a
later commit and add some tests there for closure capture inference.

dianne committed:
refactor struct pattern checking to get info for peeling

See the previous commit for details. This doesn't yet extract the struct
pat's type's ADT def before peeling, but it should now be possible.

dianne committed:
refactor tuple struct pattern checking to get info for peeling

See the previous two commits.

dianne committed:
pattern typing for immutable implicit deref patterns

dianne committed:
don't peel ADTs the pattern could match

This is the use for the previous commits' refactors; see the messages
there for more information.

dianne committed:
register `DerefMut` bounds for implicit mutable derefs

dianne committed:
respect the tcx's recursion limit when peeling

dianne committed:
upvar inference for implicit deref patterns

dianne committed:
add a feature gate test

Implicit deref patterns allow previously ill-typed programs. Make sure
they're still ill-typed without the feature gate. I've thrown in a test
for `deref!(_)` too, though it seems it refers to `deref_patterns` as a
library feature.

dianne committed:
add an unstable book chapter

- Clarifies the uses of implicit and explicit deref patterns
- Includes motivating examples for both syntaxes
- Shows the interaction with match ergonomics for reference types
- Cross-links with `string_deref_patterns` and `box_patterns`

The examples are contrived, but hopefully the intent comes across.

dianne committed:
store the kind of pattern adjustments in `pat_adjustments`

This allows us to better distinguish builtin and overloaded implicit
dereferences.

dianne commented on tests/ui/pattern/deref-patterns/typeck_fail.stderr:
The implicit and explicit deref patterns report different expected types when matching on a string since string literal patterns are `AdjustMode::Pass` (meaning the `String` doesn't get peeled either). As a fun bonus, I think that means this is compatible with the `string_deref_patterns` feature, if we don't end up going with a more general solution.

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -501,7 +568,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             PatKind::Expr(lt) => match self.resolve_vars_if_possible(self.check_pat_expr_unadjusted(lt)).kind() {
                 ty::Ref(..) => AdjustMode::Pass,
-                _ => AdjustMode::Peel,
+                // If an inline const pattern has a library-defined pointer-like type and
+                // `deref_patterns` is enabled, don't implicitly add deref patterns for its ADT.
+                // Nothing in the library that implements `DerefPure` supports structural equality,
+                // but we don't check that until `const_to_pat` in THIR construction. By avoiding
+                // type errors here, we can get a more meaningful error there.
+                ty::Adt(adt, _) => AdjustMode::peel_until_adt(Some(*adt)),
```
I think this'll be dead if #138492 lands: no other expression patterns can be `ty::Adt(..)`, right? I don't love leaving untested code in, so maybe this case should be removed (and maybe replaced with a debug assert?) if so. It's not too catastrophic if that assumption ends up being wrong: it'd lead to confusing diagnostics for a relatively niche error.

Nadrieril replied:
+1 on making this a debug assert

dianne replied:
Done [here](https://github.com/rust-lang/rust/compare/fc51373494805e0e2ead0d83bb8e66b9d49aad87..3b91b7ac57a534bbb1d4f39f679f86040a588903?diff=unified&w=#diff-9f26e4ec8d6ac64edbb3532a590592556b268b0e33f9fd4264d5d449aebbecf7) at the bottom of the `pat.rs` diff (I can't seem to link to an individual line). Ideally I'd like if the way expression patterns were handled was a bit more consistent, but that'd be a job for a separate PR.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13869124379/job/38813464809) [(plain)](https://github.com/rust-lang/rust/commit/be3391126cb76bc1f93324b9d6cafd6c2aded3a9/checks/38813464809/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#20 exporting to docker image format
#20 sending tarball 19.5s done
#20 DONE 24.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
---
      |                         ^
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/libgccjit.cc:30:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/libgccjit.cc: In function ‘gcc_jit_type* gcc_jit_context_new_array_type(gcc_jit_context*, gcc_jit_location*, gcc_jit_type*, long unsigned int)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/libgccjit.cc:807:37: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
  807 |   RETURN_NULL_IF_FAIL (num_elements >= 0, ctxt, NULL, "negative size");
---
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.h:31,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/dummy-frontend.cc:25:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/../libgcc/libgcov-util.c: In function ‘gcov_info* gcov_read_profile_dir(const char*, int)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/../libgcc/libgcov-util.c:456:9: warning: ignoring return value of ‘int chdir(const char*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  456 |   chdir (pwd);
      |   ~~~~~~^~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.h:31,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.cc:49:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.h:31,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:24:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def: At global scope:
./options.h:6899:37: warning: narrowing conversion of ‘global_options.gcc_options::x_flag_openacc’ from ‘int’ to ‘bool’ [-Wnarrowing]
 6899 | #define flag_openacc global_options.x_flag_openacc
      |                      ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:58:23: note: in definition of macro ‘DEF_BUILTIN’
   58 |   {NAME, CLASS, TYPE, BOTH_P, FALLBACK_P, ATTRS, IMPLICIT},
      |                       ^~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/builtins.def:223:16: note: in expansion of macro ‘flag_openacc’
  223 |                flag_openacc, true, true, ATTRS, false, true)
      |                ^~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def:55:1: note: in expansion of macro ‘DEF_GOACC_BUILTIN_COMPILER’
   55 | DEF_GOACC_BUILTIN_COMPILER (BUILT_IN_ACC_ON_DEVICE, "acc_on_device",
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
./options.h:6911:36: warning: narrowing conversion of ‘global_options.gcc_options::x_flag_openmp’ from ‘int’ to ‘bool’ [-Wnarrowing]
 6911 | #define flag_openmp global_options.x_flag_openmp
      |                     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:58:23: note: in definition of macro ‘DEF_BUILTIN’
   58 |   {NAME, CLASS, TYPE, BOTH_P, FALLBACK_P, ATTRS, IMPLICIT},
      |                       ^~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/builtins.def:238:16: note: in expansion of macro ‘flag_openmp’
  238 |                flag_openmp, true, true, ATTRS, false, flag_openmp)
      |                ^~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def:72:1: note: in expansion of macro ‘DEF_GOMP_BUILTIN_COMPILER’
   72 | DEF_GOMP_BUILTIN_COMPILER (BUILT_IN_OMP_IS_INITIAL_DEVICE,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In function ‘long unsigned int get_random_number()’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:10688:12: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
10688 |       read (fd, &ret, sizeof (HOST_WIDE_INT));
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gengtype-lex.l: In function ‘int yylex(const char**)’:
gengtype-lex.cc:357:15: warning: this statement may fall through [-Wimplicit-fallthrough=]
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc: In member function ‘virtual void gcc::jit::recording::memento_of_set_personality_function::replay_into(gcc::jit::replayer*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:7595:72: warning: unused parameter ‘r’ [-Wunused-parameter]
 7595 | recording::memento_of_set_personality_function::replay_into (replayer *r)
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.300
[RUSTC-TIMING] test test:false 8.893
    Finished `release` profile [optimized] target(s) in 1m 20s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 80.572
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.816
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 3.026
    Finished `release` profile [optimized] target(s) in 4.19s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.198
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.186
---
---- compiler/rustc_middle/src/ty/typeck_results.rs - ty::typeck_results::TypeckResults::pat_adjustments (line 91) stdout ----
error[E0308]: mismatched types
 --> compiler/rustc_middle/src/ty/typeck_results.rs:93:5
  |
3 | match &Box::new(Some(5i32)) {
  |       --------------------- this expression has type `&Box<Option<i32>>`
4 |     Some(n) => {},
  |     ^^^^^^^ expected `Box<Option<i32>>`, found `Option<_>`
  |
  = note: expected struct `Box<Option<i32>>`
               found enum `Option<_>`
help: consider dereferencing to access the inner value using the Deref trait
  |
3 | match &*Box::new(Some(5i32)) {
  |        +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13869522116/job/38814374630) [(plain)](https://github.com/rust-lang/rust/commit/dc134ca89e7422f1e68504ad7773c9e9d7921b77/checks/38814374630/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#20 exporting to docker image format
#20 sending tarball 20.0s done
#20 DONE 26.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
  640 |     if ((is_int () && other->is_int () || is_float() && other->is_float())
      |          ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def: At global scope:
./options.h:6899:37: warning: narrowing conversion of ‘global_options.gcc_options::x_flag_openacc’ from ‘int’ to ‘bool’ [-Wnarrowing]
 6899 | #define flag_openacc global_options.x_flag_openacc
      |                      ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:58:23: note: in definition of macro ‘DEF_BUILTIN’
   58 |   {NAME, CLASS, TYPE, BOTH_P, FALLBACK_P, ATTRS, IMPLICIT},
      |                       ^~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/builtins.def:223:16: note: in expansion of macro ‘flag_openacc’
  223 |                flag_openacc, true, true, ATTRS, false, true)
      |                ^~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def:55:1: note: in expansion of macro ‘DEF_GOACC_BUILTIN_COMPILER’
   55 | DEF_GOACC_BUILTIN_COMPILER (BUILT_IN_ACC_ON_DEVICE, "acc_on_device",
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
./options.h:6911:36: warning: narrowing conversion of ‘global_options.gcc_options::x_flag_openmp’ from ‘int’ to ‘bool’ [-Wnarrowing]
 6911 | #define flag_openmp global_options.x_flag_openmp
      |                     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:58:23: note: in definition of macro ‘DEF_BUILTIN’
   58 |   {NAME, CLASS, TYPE, BOTH_P, FALLBACK_P, ATTRS, IMPLICIT},
      |                       ^~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/builtins.def:238:16: note: in expansion of macro ‘flag_openmp’
  238 |                flag_openmp, true, true, ATTRS, false, flag_openmp)
      |                ^~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/omp-builtins.def:72:1: note: in expansion of macro ‘DEF_GOMP_BUILTIN_COMPILER’
   72 | DEF_GOMP_BUILTIN_COMPILER (BUILT_IN_OMP_IS_INITIAL_DEVICE,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In function ‘long unsigned int get_random_number()’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:10688:12: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
10688 |       read (fd, &ret, sizeof (HOST_WIDE_INT));
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-everything’ may have been intended to silence earlier diagnostics
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gengtype-lex.l: In function ‘int yylex(const char**)’:
gengtype-lex.cc:357:15: warning: this statement may fall through [-Wimplicit-fallthrough=]
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
At global scope:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.305
[RUSTC-TIMING] test test:false 8.952
    Finished `release` profile [optimized] target(s) in 1m 21s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 81.547
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.820
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 3.049
    Finished `release` profile [optimized] target(s) in 4.35s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.200
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.188
---
---- compiler/rustc_middle/src/ty/typeck_results.rs - ty::typeck_results::TypeckResults::pat_adjustments (line 91) stdout ----
error[E0308]: mismatched types
 --> compiler/rustc_middle/src/ty/typeck_results.rs:93:5
  |
3 | match &Box::new(Some(5i32)) {
  |       --------------------- this expression has type `&Box<Option<i32>>`
4 |     Some(n) => {},
  |     ^^^^^^^ expected `Box<Option<i32>>`, found `Option<_>`
  |
  = note: expected struct `Box<Option<i32>>`
               found enum `Option<_>`
help: consider dereferencing to access the inner value using the Deref trait
  |
3 | match &*Box::new(Some(5i32)) {
  |        +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
```

</details>

dianne commented on compiler/rustc_hir_typeck/src/expr_use_visitor.rs:
```diff
@@ -1726,12 +1728,31 @@ impl<'tcx, Cx: TypeInformationCtxt<'tcx>, D: Delegate<'tcx>> ExprUseVisitor<'tcx
+                    // This adjustment corresponds to an overloaded deref; it borrows the scrutinee to
+                    // call `Deref::deref` or `DerefMut::deref_mut`. Invoke the callback before setting
+                    // `place_with_id` to the temporary storing the result of the deref.
+                    // HACK(dianne): giving the callback a fake deref pattern makes sure it behaves the
+                    // same as it would if this were an explicit deref pattern.
+                    op(&place_with_id, &hir::Pat { kind: PatKind::Deref(pat), ..*pat })?;
```
I give some alternatives to this hack in the commit message for https://github.com/rust-lang/rust/pull/138528/commits/0da041b5f26ef2693ddba34f7a51ba595c566493. Following #138961, `maybe_read_scrutinee` won't exist, so if that lands, I think the way to go might be inlining this `op` call. Maybe later it'd make sense to refactor `walk_pat`/`cat_pattern` to get rid of the callback altogether? I don't think the indirection makes the code any easier to read. Though there's a lot of refactoring that could be done at that point.. with `maybe_read_scrutinee` gone, I don't think anything done after hitting a deref pattern actually matters; no uses of `ExprUseVisitor` care about how temporaries are borrowed.

dianne commented:
Rebased to resolve a conflict: as of #138492, the `inline_const_pat` feature gate has been removed, so I've removed[^1] a test for the interaction between implicit deref pattern typing and that feature. The implementation of inline const block patterns has been left intact though, so for now I've left in the inline-const-specific case in typeck that the test was for.

[^1]: I added a `// FIXME(inline_const_pat)` comment to the test file in its place, in case `inline_const_pat` is eventually reinstated.

Nadrieril commented on compiler/rustc_middle/src/ty/adjustment.rs:
```diff
@@ -214,3 +214,23 @@ pub enum CustomCoerceUnsized {
     Struct(FieldIdx),
 }
+
+/// Represents implicit coercions of patterns' types, rather than values' types.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PatAdjust {
+    /// An implicit dereference before matching, such as when matching the pattern `0` against a
+    /// scrutinee of type `&u8` or `&mut u8`.
+    // FIXME(deref_patterns): this can also be used for boxes to allow moving out of them
+    BuiltinDeref,
+    /// An implicit call to `Deref(Mut)::deref(_mut)` before matching, such as when matching the
+    /// pattern `[..]` against a scrutinee of type `Vec<T>`.
+    OverloadedDeref,
+}
```
I would really like that the `pat_adjustments` list directly stores this (plus the type), instead of using the type as an indirect way to remember what kind of deref happened.

dianne replied:
done: https://github.com/rust-lang/rust/pull/138528/commits/64c791702841551016a05fc233705c2b57611c16. ideally I'd like for this to clarify preexisting uses of `pat_adjustments` too

Nadrieril commented on compiler/rustc_hir_typeck/src/expr_use_visitor.rs:
```diff
@@ -1674,12 +1676,30 @@ impl<'tcx, Cx: TypeInformationCtxt<'tcx>, D: Delegate<'tcx>> ExprUseVisitor<'tcx
+                    // At the end of the deref chain, we get `pat`'s scrutinee.
+                    None => self.pat_ty_unadjusted(pat)?,
+                };
+                self.pat_deref_temp(pat.hir_id, pat, target_ty)?
```
Should we `break` here?

dianne replied:
I think we could, but we'd have to be careful about it. If we keep traversing the pattern without updating `place_with_id`, I think we'd introduce an ICE similar to #125059, since we expect `place_with_id.place.ty()` to match the scrutinee's type. Some options:
- Once `maybe_read_scrutinee` is gone, I think we won't have to visit the subpatterns of overloaded deref patterns at all (either implicit or explicit); no uses of `ExprUseVisitor` care about how temporary rvalues are used. i.e. we could `return` both there and in the case for `PatKind::Deref` below.
- We could also stop traversing there before `maybe_read_scrutinee` is gone, anticipating its removal. This would bring in some of the breaking change from #138961, but it shouldn't actually break anything here yet since this is gated behind an unstable feature.
- If there's some reason to keep traversing but we want to break that loop early, we could cheat a bit and skip to making a temporary with the type expected by the pattern (this also depends on `ExprUseVisitor` consumers not caring about how temporaries are used):
```rust
let target_ty = self.pat_ty_unadjusted(pat)?;
place_with_id = self.pat_deref_temp(pat.hir_id, pat, target_ty)?;
break;
```

Nadrieril replied:
Seems fine to leave it, we'll see how https://github.com/rust-lang/rust/pull/138961 develops.

Nadrieril reviewed (changes requested):
About the middle 4 commits (the ones that set up `ResolvedPat`): I think we can do simpler. To start with, https://github.com/rust-lang/rust/pull/139449 should make peeling simpler overall, and then I'd prefer an enum over `dyn Fn` callbacks (cf Zulip DMs).

About the rest of the commits, LGTM and I left some comments.

mejrs commented on src/doc/unstable-book/src/language-features/deref-patterns.md:
```diff
@@ -0,0 +1,48 @@
+
+pub fn main() {
+    let mut v = vec![Box::new(Some(0))];
+    if let [Some(ref mut x)] = v {
+        *x += 1;
+    }
```
It's best to avoid the ref mut here and elsewhere (if possible?)

```suggestion
    if let [Some(x)] = &mut v {
        *x += 1;
    }
```

mejrs commented on src/doc/unstable-book/src/language-features/deref-patterns.md:
```diff
@@ -0,0 +1,48 @@
+------------------------
+
+This feature permits pattern matching on [library-defined smart pointers] through their `Deref`
+target types, either implicitly or with the placeholder `deref!(_)` syntax.
```
From the hackmd it looks like the `deref!` syntax is required in some cases - an example of that would be nice.

Nadrieril replied:
[This](https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg) hackmd? Where do you see that?

mejrs replied:
At https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg#Explicit-vs-implicit-syntax

> Moreover, we need explicit syntax to disambiguate cases like:
> ```rust
> let cowcow: Cow<Cow<bool>> = ...;
> match cowcow {
>    // How do I match the inner cow?
>    Cow::Borrowed(_) => ...,
>    _ => ...,
>}
> ```

Has the current implementation diverged from this?

Nadrieril replied:
I see. I don't think this particular edge case belongs in the unstable guide, but an example with explicit `deref!` would make sense.

dianne replied:
I've added an example that requires explicit `deref!` [here](https://github.com/rust-lang/rust/pull/138528/commits/a006927a2e0c6029d424013e676852de39831543). I had a bit of trouble coming up with an example that was both concise and couldn't be done more simply in the body of the match arm, so it's unfortunately pretty contrived. Any ideas for how to make it make more sense?

mejrs replied:
"when finer control is needed" is good .If we think of or come across a more compelling example we can add it later 👍

mejrs commented on src/doc/unstable-book/src/language-features/deref-patterns.md:
```diff
@@ -0,0 +1,48 @@
+
+------------------------
+
+This feature permits pattern matching on [library-defined smart pointers] through their `Deref`
```
How does this compose with `string_deref_patterns`? Can you make this page link to `./string-deref-patterns.html` and back?

dianne replied:
I've added the links [here](https://github.com/rust-lang/rust/pull/138528/commits/a006927a2e0c6029d424013e676852de39831543), plus some clarification on how the features are related. I'm not sure about the wording though.

mejrs replied:
"when finer control is needed" is good .If we think of or come across a more compelling example we can add it later 👍

dianne replied:
Ah, ty! That does look nicer. I've applied a similar treatment to the `box_patterns` page, since my understanding is `deref_patterns` will replace that too (judging by the discussion on #29641). Revised commit: https://github.com/rust-lang/rust/pull/138528/commits/d412694f8fa00078905a5f493b130da11af90b69

mejrs reviewed (commented):
Thank you so much for pushing this feature! I'm really looking forward to using this. I can't really comment on the implementation but have some thoughts about the unstable book entry.

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -161,15 +161,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+/// Restrictions on what types to peel when adjusting the expected type and binding mode.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum PeelKind {
+    /// Only peel reference types. This is used for explicit deref patterns.
```
It is about `deref!()` patterns: the current behavior is that `deref!()` will peel any number of `&`/`&mut`, but stop at the first non-ref type. I'll clarify the comment if the variant stays around, but I think it might not be necessary depending on how we approach peeling. e.g. if we invert the logic so we figure out how to type the pattern before we try peeling, `deref!()`'s peeling behavior could be implemented in terms of whatever mechanism is used to do that. Or possibly `deref!()` should be `AdjustMode::Pass` for better consistency?

Nadrieril replied:
oh! a more detailed comment would be helpful yes :D

dianne replied:
done here: https://github.com/rust-lang/rust/pull/138528/commits/2933772fa64bd1a7ecae41c82441c5ee22a54633. I also changed the name of `RefOnly` to `ExplicitDerefPat`. the asymmetry doesn't feel quite right to me, so maybe some other name would be better, but I'd like it to reflect the intended mental model as well

dianne commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -161,15 +161,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    /// for `until_adt` if the pattern is a constructor. Otherwise this is the same as `RefOnly`.
+    /// See [`ResolvedPat`] for more information.
+    // TODO: add `ResolvedPat` and `until_adt`.
+    Overloaded,
```
That might be better, yeah[^1]. Is the mental model in that case that `deref!()` explicitly dereferences not only the smart pointer it matches, but also all the `&`s and `&mut`s that get peeled in order to get there?

[^1]: if that enum stays; see [my reply about `RefOnly`](https://github.com/rust-lang/rust/pull/138528/files/4c3ead8dc0acdebe79abb2e1dcbcf458a0fd55db..bf722f148d2cb321716fd42ecacc156c4e777343#r2030212020)

Nadrieril replied:
That's my mental model yes

dianne committed:
refactor path pattern checking to get info for peeling

See the doc comment on `ResolvedPat` for more information. This and the
next couple commits split resolution apart from checking for path,
struct, and tuple struct patterns, in order to find the pattern's type
before peeling the scrutinee. This helps us avoid peeling the scrutinee
when the pattern could match it.

The reason this handles errors from resolution after peeling is for
struct and tuple struct patterns: we check their subpatterns even when
they fail to resolve, to potentially catch more resolution errors. By
doing this after peeling, we're able to use the updated `PatInfo`. I
don't know if there's currently any observable difference from using the
outdated `PatInfo`, but it could potentially be a source of subtle
diagnostic bugs in the future, so I'm opting to peel first.

dianne commented:
Ty for the reviews! I've pushed an update to hopefully address them. Most of the changes I've added as new commits for ease of comparison, but since the `ResolvedPat` refactoring was kind of severe, I figured it'd be cleanest to edit the commit history so that it can be compared more easily with the `master` branch. In case it helps with the review, I have the old commit history archived [here](https://github.com/dianne/rust/commits/old-implicit-deref-patterns/). I haven't figured out how to get a diff-of-diffs on github.

As for `ResolvedPat` and `PeelKind`, I'm still not totally happy with them, but I've been struggling to come up with anything simpler.

I'm not marking this as ready for review yet, since I haven't rebased on top of #139449 yet.

dianne commented on compiler/rustc_mir_build/src/thir/pattern/mod.rs:
```diff
@@ -66,9 +66,11 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
             self.typeck_results.pat_adjustments().get(pat.hir_id).map_or(&[], |v| &**v);
 
         // Track the default binding mode for the Rust 2024 migration suggestion.
+        // Implicitly dereferencing references changes the default binding mode, but implicit deref
+        // patterns do not. Only track binding mode changes if a ref type is in the adjustments.
```
done in https://github.com/rust-lang/rust/commit/2933772fa64bd1a7ecae41c82441c5ee22a54633, with a couple extra wording changes to hopefully make it make more sense in context

mejrs commented on src/doc/unstable-book/src/language-features/deref-patterns.md:
```diff
@@ -0,0 +1,61 @@
+#![feature(deref_patterns)]
+#![allow(incomplete_features)]
+
+pub fn main() {
```
The main function is not necessary (both here and below), mdbook will insert one. You can also hide it with `# fn main(){` if you hit one of the corner cases where it is needed.

dianne replied:
I totally forgot about that, ty! Applied and squashed: https://github.com/rust-lang/rust/pull/138528/commits/94295d5fd8c090b4391bb2541fe4acb1e97b8080

mejrs reviewed (approved):
I like this iteration of the docs, just one nit..

bors commented:
:umbrella: The latest upstream changes (presumably #139622) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne commented:
I cleaned up the history a bit while rebasing. The change to make the `pat_adjustments` table contain the kind of each implicit deref is now the first commit, so everything else can properly depend on it. @rustbot ready

dianne commented on compiler/rustc_middle/src/ty/adjustment.rs:
```diff
@@ -214,3 +214,26 @@ pub enum CustomCoerceUnsized {
+pub enum PatAdjust {
+    /// An implicit dereference before matching, such as when matching the pattern `0` against a
+    /// scrutinee of type `&u8` or `&mut u8`.
+    // FIXME(deref_patterns): this can also be used for boxes to allow moving out of them
+    BuiltinDeref,
```
After thinking about this more and looking into how to implement exhaustiveness checking, I'm leaning towards not using this for boxes. My plan was to lower deref patterns on boxes to `thir::PatKind::Deref`, similar to the current implementation of box patterns. However, I'm now thinking deref patterns on boxes should be lowered to `thir::PatKind::DerefPattern` (same as other deref patterns), which can get special treatment for boxes in `MatchPairTree::for_pattern`. Either approach is about the same amount of work to implement (both are very small changes), except:
1. Using `PatAdjust::OverloadedDeref` for boxes makes handling implicit and explicit deref patterns a bit more consistent.
2. Using `thir::PatKind::DerefPattern` for boxes means we should be able to get rid of all the special-casing for boxes in exhaustiveness checking, once box patterns are properly replaced by (or implemented in terms of) deref patterns.

In any case, I'll push an update to remove the box-specific FIXMEs. Regardless of whether using `thir::PatKind::Deref` for boxes is preferable, I have it implemented both ways locally already, so the reminders aren't necessary.

Nadrieril replied:
Agreed, treating boxes like a normal smart pointer seems overall cleaner.

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/migration.rs:
```diff
@@ -93,10 +93,10 @@ impl<'a> PatMigration<'a> {
-        let implicit_deref_mutbls = adjustments.iter().map(|ref_ty| {
-            let &ty::Ref(_, _, mutbl) = ref_ty.kind() else {
+        let implicit_deref_mutbls = adjustments.iter().map(|adjust| {
+            let &ty::Ref(_, _, mutbl) = adjust.source.kind() else {
```
In a future PR, we could store the mutability here, which would remove the need for matching on the type here and in clippy.

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/mod.rs:
```diff
@@ -68,9 +68,11 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
         // Track the default binding mode for the Rust 2024 migration suggestion.
+        // Implicitly dereferencing references changes the default binding mode, but implicit deref
+        // patterns do not. Only track binding mode changes if a ref type is in the adjustments.
         let mut opt_old_mode_span = None;
         if let Some(s) = &mut self.rust_2024_migration
-            && !adjustments.is_empty()
+            && adjustments.iter().any(|adjust| adjust.source.is_ref())
```
I would prefer to match on the `PatAdjust` here.

dianne replied:
done [here](https://github.com/rust-lang/rust/compare/fc51373494805e0e2ead0d83bb8e66b9d49aad87..3b91b7ac57a534bbb1d4f39f679f86040a588903?diff=unified&w=#diff-d83cf17741a9ac1741c5ce5e5fe4e3cdce97cd7e47f775086d7bdac9a341aef0). that's definitely better with `PatAdjust::BuiltinDeref` only being for `ty::Ref(..)`, yeah

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -161,26 +161,34 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// don't peel it. See [`ResolvedPat`] for more information.
-    // TODO: add `ResolvedPat` and `until_adt`.
-    Implicit,
+    Implicit { until_adt: Option<AdtDef<'tcx>> },
```
nit: `AdtDef` has too much info, I think this should just be the adt's `DefId`.

dianne replied:
done [here](https://github.com/rust-lang/rust/compare/fc51373494805e0e2ead0d83bb8e66b9d49aad87..3b91b7ac57a534bbb1d4f39f679f86040a588903?diff=unified&w=#diff-9f26e4ec8d6ac64edbb3532a590592556b268b0e33f9fd4264d5d449aebbecf7)

Nadrieril commented on tests/ui/pattern/deref-patterns/implicit-cow-deref.rs:
```diff
@@ -0,0 +1,44 @@
+use std::borrow::Cow;
+
+fn main() {
+    let cow: Cow<'static, [u8]> = Cow::from(&[1, 2, 3]);
```
I had to look at the docs to check which one that was x)
```suggestion
    let cow: Cow<'static, [u8]> = Cow::Borrowed(&[1, 2, 3]);
```

Nadrieril commented on tests/ui/pattern/deref-patterns/implicit-cow-deref.rs:
```diff
@@ -0,0 +1,44 @@
+        _ => unreachable!(),
+    }
+
+    match cow_of_cow {
```
Worth a comment for future readers.
```suggestion
    // This matches on the outer `Cow` (the owned one).
    match cow_of_cow {
```

Nadrieril committed:
Make cow_of_cow test a teeny bit more explicit

Nadrieril reviewed (approved):
Looking gooood! I'm very happy with this PR now. I left a few nits that aren't too important, except I agree with your proposal [here](https://github.com/rust-lang/rust/pull/138528#discussion_r1996557298) to check if there can be an expression pattern without inline_const_pats. I'm in favor of doing it in this PR so we don't forget.

Nadrieril commented:
let's goo

@bors r+

bors commented:
:pushpin: Commit 3b91b7ac57a534bbb1d4f39f679f86040a588903 has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 3b91b7ac57a534bbb1d4f39f679f86040a588903 -->
<!-- homu: {"type":"Approved","sha":"3b91b7ac57a534bbb1d4f39f679f86040a588903","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
